core/jvm/src/main/scala/zio/prototype.scala

```diff
@@ -0,0 +1,130 @@
+
+    final def update[A](lens: Lens[Environment, A])(f: A => A): UIO[Unit] =
+      modify(lens)(a => ((), f(a)))
+
+    final def use[A, B](lens: Lens[Environment, A])(f: A => B): UIO[B] =
+      modify(lens)(a => (f(a), a))
+
+    final def useM[A, B](lens: Lens[Environment, A])(f: A => UIO[B]): UIO[B] =
+      modify(lens)(a => (f(a), a)).flatten
+  }
+
+  object EnvironmentalEffect {
+
+    def modify[R <: EnvironmentalEffect[Environment], Environment, A, B](
+      lens: Lens[Environment, A]
+    )(f: A => (B, A)): ZIO[R, Nothing, B] =
+      ZIO.accessM[R](_.modify(lens)(f))
+
+    def update[R <: EnvironmentalEffect[Environment], Environment, A](
+      lens: Lens[Environment, A]
+    )(f: A => A): ZIO[R, Nothing, Unit] =
+      ZIO.accessM[R](_.update(lens)(f))
+  }
+
+  trait Logging {
+    def logging: Logging.Service
+  }
+
+  object Logging {
+    trait Service {
+      def log(line: String): UIO[Unit]
+    }
+  }
+
+  trait LoggingEffect[R] extends EnvironmentalEffect[R] {
+    def logging: Lens[R, Logging.Service]
+  }
+
+  object LoggingEffect {
+
+    trait Live extends LoggingEffect[ZEnv with Logging] {
+
+      val environment: Ref[ZEnv with Logging]
+
+      def logging: Lens[ZEnv with Logging, Logging.Service] =
+        (
+          s => s.logging,
+          a =>
+            s =>
+              new Clock with Console with System with Random with Blocking with Logging {
+                val clock    = s.clock
+                val console  = s.console
+                val system   = s.system
+                val random   = s.random
+                val blocking = s.blocking
+                val logging  = a
+              }
+        )
+
+      def modify[A, B](lens: Lens[ZEnv with Logging, A])(f: A => (B, A)): UIO[B] =
+        environment.get.flatMap { r =>
+          val a       = lens._1(r)
+          val (b, a1) = f(a)
+          environment.set(lens._2(a1)(r)) *> ZIO.succeed(b)
+        }
+    }
+
+    def make: ZIO[ZEnv, Nothing, LoggingEffect[ZEnv with Logging]] =
+      ZIO.accessM[ZEnv] { r =>
+        val env: ZEnv with Logging = new Clock with Console with System with Random with Blocking with Logging {
+          val clock    = r.clock
+          val console  = r.console
+          val system   = r.system
+          val random   = r.random
+          val blocking = r.blocking
+          val logging = new Logging.Service {
+            def log(line: String): UIO[Unit] = console.putStrLn(line)
+          }
+        }
+        Ref.make(env).map { ref =>
+          new LoggingEffect.Live {
+            val environment: Ref[ZEnv with Logging] = ref
+          }
+        }
+      }
+
+    def log(line: String): ZIO[LoggingEffect[ZEnv with Logging], Nothing, Unit] =
+      ZIO.accessM[LoggingEffect[ZEnv with Logging]](service => service.useM(service.logging)(_.log(line)))
+
+    def addPrefix(prefix: String): ZIO[LoggingEffect[ZEnv with Logging], Nothing, Unit] =
```

Comment:
I'd make this polymorphic: `addPrefix[R <: Logging]` to show that `addPrefix` doesn't know what the environment type is.

Reply:
Yes, done now.